hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java

```diff
@@ -58,6 +58,31 @@ default T preCombine(T oldValue, Properties properties) {
     return preCombine(oldValue);
   }
 
+  /**
+   *When more than one HoodieRecord have the same HoodieKey in the incoming batch, this function combines them before attempting to insert/upsert by taking in a property map.
+   *
+   * @param oldValue instance of the old {@link HoodieRecordPayload} to be combined with.
+   * @param properties Payload related properties. For example pass the ordering field(s) name to extract from value in storage.
+   * @param schema Schema used for record
+   * @return the combined value
+   */
+  @PublicAPIMethod(maturity = ApiMaturityLevel.STABLE)
+  default T preCombine(T oldValue, Properties properties, Schema schema) {
```

Comment:
Thanks @alexeykudinkin for reviewing this.

What we are trying to do is implement partial update purpose. For example, let's assume the record schema is `(f0 int , f1 int, f2 int)`, The first record value is: (1, 2, 3), the second record value is: (4, 5, null) with the field f2 value as null. We hope that the result after run `preCombine` is `(4, 5, 3)`, which means we need to combine/merge two records to a third one, not only choose one of them.

Actually, what we want to implement is similar with `#combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema)` which used for combine the incoming record with existing record from base/log file.
But `#preCombine` will be used for combing/merging two incoming records in a batch.

Reply:
Right, that's exactly my question: why do you want to implement such semantic w/in `preCombine`? What use-case you're trying to accommodate for here? 

Essentially with this change you will introduce a way for 2 records w/in the batch to be combined into 1. But why do you need this? 

After all you can achieve the same goal if you just stop de-duping your records, and then subsequently merge them against what is on disk

Reply:
Hi @alexeykudinkin, I got your point. if we have to combine two records to a combined one, we'd better to implement the combine logics in other place, maybe in some `util` or `helper` classes, right?

Here are some options from mine that `#preCombine` might be a better place to implement these logics, or create new `merge` method in `HoodieRecordPayload`  interface.
- First, from the description of `preCombine` method, it used for combining multiple records with same HoodieKey before attempting to insert/upsert to disk. The "combine multiple records" might not mean only choosing one of them, we also can combine & merged them to a new one, just depends on how the sub-class implement the preCombine logic(Please correct me if my understanding is wrong :) ). Yeah, it might be a little bit confused that we need `Schema` if we are trying to merged them.
- Second, I checked when will we call `preCombine` method is trying to duplicate records with same HoodieKey before insert/update to disk, especially in Flink write case, even through the duplicated logic is choose the latest record, but we need to ensure that one HoodieKey should only contains one record before comparing to existing record and write to disk, otherwise, some records will missed. For example, in `HoodieMergeHandle.init(fieId, newRecordsIter)`, it will convert the record iterator to a map and treat the recordKey as key. *So we might not stop de-duping logics and merge them against what is on disk unless we change the logic here.* And also we implement another class/method to handle the merge logic, and switch the existing de-duping logic from calling `preCombine` to new class/method, we have to add an condition to control whether should we call `preCombine` or not, I think it might not a good way. Instead, we should handle it in `preCombine` method by different implemented payload.

That's what my thought here, and I'm glad to listen your useful suggestions. :)

Reply:
Let me try to clarify a few things: 

`preCombine` has a _very specific_ semantic: it's de-duplicating by the way of picking "most recent" among records in the batch. Expectation always is that it being handed 2 records it will **have to** return either of them. It could not produce new record. If we want to revisit this semantic this is a far larger change that will surely require writing an RFC and broader discussion regarding the merits of such migration. Please also keep in mind that as of RFC-46 there's an effort underway to abstract whole "record combination/merging" semantic out of `RecordPayload` hierarchy into standalone Combination/Merge Engine API.

> First, from the description of preCombine method, it used for combining multiple records with same HoodieKey before attempting to insert/upsert to disk. The "combine multiple records" might not mean only choosing one of them, we also can combine & merged them to a new one, just depends on how the sub-class implement the preCombine logic(Please correct me if my understanding is wrong :) ). Yeah, it might be a little bit confused that we need Schema if we are trying to merged them.

Please see my comment regarding `preCombine` semantic above. I certainly agree with you that the name is confusing, but i've tried to clear that confusion. Let me know if you have more questions about it.

> Second, I checked when will we call preCombine method is trying to duplicate records with same HoodieKey before insert/update to disk, especially in Flink write case, even through the duplicated logic is choose the latest record, but we need to ensure that one HoodieKey should only contains one record before comparing to existing record and write to disk, otherwise, some records will missed. For example, in HoodieMergeHandle.init(fieId, newRecordsIter), it will convert the record iterator to a map and treat the recordKey as key. So we might not stop de-duping logics and merge them against what is on disk unless we change the logic here. And also we implement another class/method to handle the merge logic, and switch the existing de-duping logic from calling preCombine to new class/method, we have to add an condition to control whether should we call preCombine or not, I think it might not a good way. Instead, we should handle it in preCombine method by different implemented payload.

You're bringing up a good points, let's dive into them one by one: so currently we have 2 mechanisms

1. `preCombine` that allows to select "most recent" record among those having the same key w/in the batch
2. `combineAndGetUpdateValue` that allows to combine previous or "historical" record (on Disk) with the new incoming one (all partial merging semantic is currently implemented in this method)

You rightfully mention some of the invariants are currently that the batch would be de-duped at certain level (b/c we have to maintain PK uniqueness on disk), and so we might need to shift that to accommodate for case that you have. And that's exactly what my question was: if you can elaborate on use-case that you have at hand that you're trying to solve w/ this PR, i would be able to better understand where you're coming from and what's the best path forward for us here.

Questions i'm looking an answers for are basically following:

1. What's nature of your use-case? (domain, record types, frequency, size, etc)
2. Where requirements for partial updates are coming from?

and etc.  I'm happy to set some 30min to talk in person regarding this or connect on Slack and discuss it there.

Reply:
Hi @alexeykudinkin , Thanks a lot for you detail clarification.
1. Regarding the design of `preCombine`, I'm clear now. I'm sorry I don't know the detail of RFC-46, and also I didn't find the link RFC-46 from [here](https://cwiki.apache.org/confluence/display/HUDI/RFC+Process), cloud you please share the link?
2. and regarding the requirements for partial updates/overwrite, I saw some same requirements from community. In my case, generally, we want to build a customer profile with multiple attributes, these attributes might come from different systems, one system might only provides some attributes in a event/record, and two systems might the events/records with different attributes, we should not only choose the recent one, we need to merged them before writing to disk. Otherwise, we have to keep all change logs, and then start a new job to dedup & merge these attributes among the change logs. For example, we have 10 attributes a1-a10(all of them are optional), source system A only has the a1-a5, source system B only has a6-a10, what result we expect is that the final record contains a1-a10, not only a1-a5 or a6-a10. And because we might receive two events/records in same time, they might be in a same batch, that's why we want to merge them before `combineAndGetUpdateValue `.

Reply:
BTW, thanks a lot for you time, will ping you on slack.

Reply:
CC @rmahindra123 who encountered a necessity to do preCombine but to combine bits and pieces from both records to return a new one. Rajesh: do you wanna go over your use-case may be.

Reply:
@alexeykudinkin , I'm sorry that I still don't find a suitable time to align with online, may i check any thoughts or suggests on this PR?